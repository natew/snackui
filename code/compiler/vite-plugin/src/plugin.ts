import type { TamaguiOptions } from '@tamagui/static'
import type { Plugin } from 'vite'
import { transformWithEsbuild } from 'vite'
import { tamaguiOptions, Static, loadTamaguiBuildConfig } from './loadTamagui'

export function tamaguiPlugin(tamaguiOptionsIn: TamaguiOptions = {}): Plugin {
  let watcher
  let loaded = false

  const extensions = [
    `.web.mjs`,
    `.web.js`,
    `.web.jsx`,
    `.web.ts`,
    `.web.tsx`,
    '.mjs',
    '.js',
    '.mts',
    '.ts',
    '.jsx',
    '.tsx',
    '.json',
  ]

  let noExternalSSR = /react-native|expo-linear-gradient/gi

  async function load() {
    loaded = true

    await loadTamaguiBuildConfig(tamaguiOptionsIn)

    if (tamaguiOptions!.disableWatchTamaguiConfig) {
      return
    }

    if (!Static) {
      throw new Error(`Not loaded`)
    }

    watcher = Static.watchTamaguiConfig({
      components: ['tamagui'],
      config: './src/tamagui.config.ts',
      ...tamaguiOptions,
    }).catch((err) => {
      console.error(` [Tamagui] Error watching config: ${err}`)
    })

    const components = [
      ...new Set([...(tamaguiOptions!.components || []), 'tamagui', '@tamagui/core']),
    ]

    noExternalSSR = new RegExp(
      `${components.join('|')}|react-native|expo-linear-gradient`,
      'ig'
    )
  }

  return {
    name: 'tamagui-base',
    enforce: 'pre',

    async buildEnd() {
      await watcher?.then((res) => {
        res?.dispose()
      })
    },

    async transform(code, id) {
      if (id.includes('expo-linear-gradient')) {
        // fix expo-linear-gradient
        return transformWithEsbuild(code, id, {
          loader: 'jsx',
          jsx: 'automatic', // ðŸ‘ˆ
        })
      }
    },

    async config(_, env) {
      await load()

      if (!tamaguiOptions) {
        throw new Error(`No options loaded`)
      }

      return {
        define: {
          // reanimated support
          _frameTimestamp: undefined,
          _WORKLET: false,
          __DEV__: `${env.mode === 'development'}`,
          'process.env.NODE_ENV': JSON.stringify(process.env.NODE_ENV || env.mode),
          'process.env.ENABLE_RSC': JSON.stringify(process.env.ENABLE_RSC || ''),
          'process.env.ENABLE_STEPS': JSON.stringify(process.env.ENABLE_STEPS || ''),
          'process.env.IS_STATIC': JSON.stringify(false),
          ...(env.mode === 'production' && {
            'process.env.TAMAGUI_OPTIMIZE_THEMES': JSON.stringify(true),
          }),
        },
        ssr: {
          noExternal: noExternalSSR,
        },
        resolve: {
          extensions: extensions,
          alias: {
            ...(tamaguiOptions.platform !== 'native' && {
              'react-native/Libraries/Renderer/shims/ReactFabric': '@tamagui/proxy-worm',
              'react-native/Libraries/Utilities/codegenNativeComponent':
                '@tamagui/proxy-worm',
              'react-native-svg': '@tamagui/react-native-svg',
              'react-native': 'react-native-web',
              ...(tamaguiOptions.useReactNativeWebLite && {
                'react-native': 'react-native-web-lite',
                'react-native-web': 'react-native-web-lite',
              }),
              ...(tamaguiOptions.useReactNativeWebLite === 'without-animated' && {
                'react-native': 'react-native-web-lite/without-animated',
                'react-native-web': 'react-native-web-lite/without-animated',
              }),
            }),
          },
        },
      }
    },
  } satisfies Plugin
}
