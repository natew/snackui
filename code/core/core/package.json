{
  "name": "@tamagui/core",
  "version": "1.115.4",
  "source": "src/index.tsx",
  "main": "dist/cjs",
  "module": "dist/esm",
  "tamagui": {
    "bundle.native": "./src/index.tsx",
    "bundle.native.test": "./src/index.tsx"
  },
  "types": "./types/index.d.ts",
  "license": "MIT",
  "author": {
    "name": "Nate Wienert"
  },
  "sideEffects": [
    "*.css"
  ],
  "scripts": {
    "build": "tamagui-build",
    "watch": "tamagui-build --watch",
    "lint": "biome check src",
    "lint:fix": "biome check --write src",
    "clean": "tamagui-build clean",
    "clean:build": "tamagui-build clean:build"
  },
  "files": [
    "types",
    "dist",
    "src",
    "reset.css",
    "native.js",
    "native.d.ts",
    "native-test.js",
    "native-test.d.ts"
  ],
  "dependencies": {
    "@tamagui/react-native-use-pressable": "1.115.4",
    "@tamagui/react-native-use-responder-events": "1.115.4",
    "@tamagui/use-event": "1.115.4",
    "@tamagui/web": "1.115.4"
  },
  "devDependencies": {
    "@tamagui/build": "1.115.4",
    "@testing-library/react": "^16.0.1",
    "csstype": "^3.0.10",
    "typescript": "^5.5.2",
    "vitest": "^2.1.3"
  },
  "exports": {
    "./package.json": "./package.json",
    "./reset.css": "./reset.css",
    ".": {
      "react-native-import": "./dist/esm/index.native.js",
      "react-native": "./dist/cjs/index.native.js",
      "types": "./types/index.d.ts",
      "import": "./dist/esm/index.mjs",
      "require": "./dist/cjs/index.cjs",
      "default": "./dist/cjs/index.native.js"
    },
    "./native": {
      "types": "./types/index.d.ts",
      "import": "./dist/native.mjs",
      "require": "./dist/native.js",
      "react-native": "./dist/native.native.js"
    },
    "./native-test": {
      "types": "./types/index.d.ts",
      "import": "./dist/test.native.js",
      "require": "./dist/test.native.js",
      "react-native": "./dist/test.native.native.js"
    },
    "./inject-styles": {
      "types": "./types/inject-styles.d.ts",
      "import": "./dist/esm/inject-styles.mjs",
      "require": "./dist/cjs/inject-styles.cjs",
      "react-native": "./dist/cjs/inject-styles.native.js"
    },
    "./reactNativeTypes": {
      "types": "./types/reactNativeTypes.d.ts",
      "import": "./dist/esm/reactNativeTypes.mjs",
      "require": "./dist/cjs/reactNativeTypes.cjs",
      "react-native": "./dist/cjs/reactNativeTypes.native.js"
    }
  },
  "repository": {
    "type": "git",
    "url": "https://github.com/tamagui/tamagui.git",
    "directory": "packages/core"
  },
  "publishConfig": {
    "access": "public"
  },
  "gitHead": "a49cc7ea6b93ba384e77a4880ae48ac4a5635c14"
}
