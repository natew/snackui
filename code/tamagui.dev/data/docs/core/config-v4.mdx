---
title: @tamagui/config
description: Easy configuration and themes with @tamagui/config/v4
---

Tamagui is flexible, but it's nice to have sensible defaults. We've been publishing those defaults under `@tamagui/config`, but until version 4 they were a bit of a "take it or leave it" affair.

Version 4 simplifies the config, and adds `createThemeSuite`, a helper function that can generate beautiful themes with just a few lines of code.
