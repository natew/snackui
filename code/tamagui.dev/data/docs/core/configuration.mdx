---
title: Configuration
description: Custom themes, tokens, shorthands, and media queries.
---

Setting up `createTamagui` is optional, but there are many useful settings so
recommend at least using a preset configuration. Create a `tamagui.config.ts`
somewhere in your app. In general, you don't want to import this file from
anywhere else but your root layout, where you'll pass it to `TamaguiProvider`.

In this guide we import everything from `tamagui`, but if you are only using the lower level
`@tamagui/core`, you'll want to import from that instead.

## Quick Start

We made `@tamagui/config` to make it easy to get started. It comes with a
recommended default configuration and themes, as well as an easy way to make
your own themes with `createThemeSuite`.

```tsx fileName="tamagui.config.ts"
import { config, defaultThemes } from '@tamagui/config/v4'
import { createTamagui } from 'tamagui' // or '@tamagui/core'

const appConfig = createTamagui([
  ...config,
  themes: defaultThemes,
])

export type AppConfig = typeof appConfig

declare module 'tamagui' { // or '@tamagui/core'
  // overrides TamaguiCustomConfig so your custom types
  // work everywhere you import `tamagui`
  interface TamaguiCustomConfig extends AppConfig {}
}

export default appConfig
```

If you'd like to create your own themes, see [the next section on @tamagui/config](/docs/config-v4).

## From scratch

Let's start with an example of a complete `tamagui.config.ts`. Note that keys
used here can be customized to your liking, and remember that they combine with
your [shorthands](#shorthands) to give you a syntax like this:

```tsx
<XStack bg="$red" w="$2" h="$1" />
```

```tsx fileName="tamagui.config.ts"
import { createFont, createTamagui, createTokens, isWeb } from 'tamagui'

// To work with the tamagui UI kit styled components (which is optional)
// you'd want the keys used for `size`, `lineHeight`, `weight` and
// `letterSpacing` to be consistent. The `createFont` function
// will fill-in any missing values if `lineHeight`, `weight` or
// `letterSpacing` are subsets of `size`.

const systemFont = createFont({
  family: isWeb ? 'Helvetica, Arial, sans-serif' : 'System',
  size: {
    1: 12,
    2: 14,
    3: 15,
  },
  lineHeight: {
    // 1 will be 22
    2: 22,
  },
  weight: {
    1: '300',
    // 2 will be 300
    3: '600',
  },
  letterSpacing: {
    1: 0,
    2: -1,
    // 3 will be -1
  },
  // (native only) swaps out fonts by face/style
  face: {
    300: { normal: 'InterLight', italic: 'InterItalic' },
    600: { normal: 'InterBold' },
  },
})

// Set up tokens

// The keys can be whatever you want, but if using `tamagui` you'll want 1-10:

const size = {
  0: 0,
  1: 5,
  2: 10,
  // ....
}

export const tokens = createTokens({
  size,
  space: { ...size, '-1': -5, '-2': -10 },
  radius: { 0: 0, 1: 3 },
  zIndex: { 0: 0, 1: 100, 2: 200 },
  color: {
    white: '#fff',
    black: '#000',
  },
})

const config = createTamagui({
  fonts: {
    heading: systemFont,
    body: systemFont,
  },
  tokens,

  themes: {
    light: {
      bg: '#f2f2f2',
      color: tokens.color.black,
    },
    dark: {
      bg: '#111',
      color: tokens.color.white,
    },
  },

  media: {
    sm: { maxWidth: 860 },
    gtSm: { minWidth: 860 + 1 },
    short: { maxHeight: 820 },
    hoverNone: { hover: 'none' },
    pointerCoarse: { pointer: 'coarse' },
  },

  // Shorthands
  // Adds <View m={10} /> to <View margin={10} />
  // See Settings section on this page to only allow shorthands
  // Be sure to have `as const` at the end
  shorthands: {
    px: 'paddingHorizontal',
    f: 'flex',
    m: 'margin',
    w: 'width',
  } as const,

  // Change the default props for any styled() component with a name.
  // We are discouraging the use of this and have deprecated it, prefer to use
  // styled() on any component to change it's styles.
  defaultProps: {
    Text: {
      color: 'green',
    },
  },
})

type AppConfig = typeof config

// this will give you types for your components
// note - if using your own design system, put the package name here instead of tamagui
declare module 'tamagui' {
  interface TamaguiCustomConfig extends AppConfig {}

  // if you want types for group styling props, define them like so:
  interface TypeOverride {
    groupNames(): 'card'
  }
}

export default config
```

The `createTamagui` function receives a configuration object:

- `tokens`: Use `createTokens` to generate variables in your theme and app.
- `theme`: Define your design [theme](/docs/intro/themes), which map to CSS
  properties.
- `media`: Define reusable responsive [media queries](/docs/core/use-media).
- `shorthands`: Define any props you want to expand to style values, keys being
  the shorthand and values being the expanded style prop.
- `defaultProps`: Deprecated, use styled() on any component to change it's
  styles. ~~For more advanced uses, you can override all named `styled()`
  components initial values.~~
- `onlyAllowShorthands`: When set to true will ensure that if you define
  shorthands, they override existing long-form style props.

<Notice>
  On Android you need to set the `face` option in `createFont` or else fonts
  won't pick up different weights, due to a React Native restriction.
</Notice>

Note, for `tamagui` (not core), it expects you to define a `true` token that
maps to your default size, this way it knows what token to use by default. So
you'd do something like this:

```tsx
export const tokens = createTokens({
  size: {
    small: 20,
    medium: 30,
    true: 30, // note true = 30 just like medium, your default size token
    large: 40,
  },
  space: {
    small: 10,
    medium: 20,
    true: 20, // same goes for space and other token categories
    large: 30,
  },
})
```

<Notice>
  The compiler will parse your tamagui.config.ts file at build-time. For this
  reason, we recommend keeping it relatively simple. Don't import heavy
  dependencies, and prefer to import `createTamagui` and other helpers directly
  from @tamagui/core.
</Notice>

### Add Provider

Import and use the Tamagui `Provider` component at the top component in your
app.

```tsx line=3-5 fileName="App.tsx"
import { TamaguiProvider } from 'tamagui'

import config from './tamagui.config'

export default function App() {
  return (
    <TamaguiProvider config={config}>
      <AppContents />
    </TamaguiProvider>
  )
}
```

#### Using Tamagui Provider

TamaguiProvider takes a few optional properties:

<PropsTable
  data={[
    {
      name: 'defaultTheme',
      required: true,
      type: 'string',
      description: `The initial top level theme.`,
    },
    {
      name: 'disableRootThemeClass',
      required: false,
      type: 'boolean',
      description:
        'Disable inserting a theme class in the DOM or context, allowing you to manually place it higher. For custom use cases like integration with next-theme.',
    },
    {
      name: 'disableInjectCSS',
      required: false,
      type: 'boolean',
      description: `By default Tamagui inserts CSS with a useInsertionEffect on load. But if you're setting up SSR you'll want to use getCSS() on the server instead and then turn on this setting.`,
    },
  ]}
/>

By default, Tamagui injects the CSS for your configuration on the client-side into `document.head`, but you
probably will want something better for production. To do this, set `disableInjectCSS`, and then:

If your framework has server-side layouts, you could just render it yourself in a root layout:

```tsx
<head>
  <style
    dangerouslySetInnerHTML={{
      __html: Tamagui.getCSS(),
    }}
  />
</head>
```

If you are using one of our bundler plugins, you can add the `outputCSS` setting, like so:

```tsx fileName=vite.config.ts
import { tamaguiPlugin } from '@tamagui/vite-plugin'

export default {
  plugins: [
    tamaguiPlugin({
      config: './src/tamagui.config.ts',
      outputCSS: './src/tamagui.css'
    }),
  ]
}
```

And then you'll want to import the `tamagui.css`.

As a final option, you can also generate it yourself with the CLI. First create a `tamagui.build.ts`:

```tsx fileName=tamagui.build.ts
import type { TamaguiBuildOptions } from 'tamagui'

export default {
  components: ['tamagui'],
  config: './config/tamagui.config.ts',
  outputCSS: './tamagui.css',
} satisfies TamaguiBuildOptions
```

And then run:

```
npx @tamagui/cli generate
```

## Tokens

Tokens have a structure inspired by but divergent from the
[Theme UI spec](https://theme-ui.com/theme-spec/). They are mapped to CSS
variables at build time.

### Font tokens

The font tokens are a bit special and are created with `createFont`:

```tsx
const interFont = createFont({
  family: 'Inter, Helvetica, Arial, sans-serif',
  size: {
    1: 12,
    2: 14,
    3: 15,
    // ...
  },
  lineHeight: {
    1: 17,
    2: 22,
    3: 25,
    // ...
  },
  weight: {
    4: '300',
    6: '600',
  },
  letterSpacing: {
    4: 0,
    8: -1,
  },

  // for native only, alternate family based on weight/style
  face: {
    // pass in weights as keys
    700: { normal: 'InterBold', italic: 'InterBold-Italic' },
    800: { normal: 'InterBold', italic: 'InterBold-Italic' },
    900: { normal: 'InterBold', italic: 'InterBold-Italic' },
  },
})
```

<Notice>
  Note, you don't need to use numbered keys, you can use `sm` or `tiny` or
  whatever you'd like. But you do want keep those keys consistent.
</Notice>

This gives you a lot of power over customizing every aspect of your design based
on each font family. In other styling libraries that follow the Theme UI spec,
you generally don't group your size/lineHeight/weight/etc tokens by the family,
which means you are forced to choose a single vertical rhythm no matter the
font.

Things to note:

- The keys of `size`, `lineHeight`, `weight`, and `letterSpacing` are meant to
  match.
- Define the full set of keys on `size`, the rest can be a subset.
- Missing keys from partially defined objects will be filled in.
  - In the example above, `weight` is only defined at `4` and `6`.
  - At creation Tamagui fills in the missing keys with previous value, or the
    next one if no previous exists. So weight `1` === `300`, weight `5` ===
    `300`, and weight `7` === `600`.

### Troubleshooting

Custom fonts not appearing in Tamagui components? Note the `family` definition,
on Native it requires a single font name, whereas on web it can include a
comma-separated list of fallbacks.

You can use the `isWeb` boolean to do this conditionally on `createFont`:

```tsx
createFont({
  family: isWeb ? 'Custom-Font, Arial, Helvetica, ...' : 'Custom-Font',
  ...rest,
})
```

#### Custom Font Family on Native

You need to load your fonts for React Native to recognize them. With Expo, this
typically looks something like this (or you can
[follow a React Native guide here](https://stackoverflow.com/questions/35255645/how-to-set-default-font-family-in-react-native)):

```tsx fileName="App.tsx"
import { useFonts } from 'expo-font'

function App() {
  const [loaded] = useFonts({
    Inter: require('@tamagui/font-inter/otf/Inter-Medium.otf'),
    InterBold: require('@tamagui/font-inter/otf/Inter-Bold.otf'),
  })

  useEffect(() => {
    if (loaded) {
      // can hide splash screen here
    }
  }, [loaded])

  if (!loaded) {
    return null
  }

  return <MyApp />
}
```

---

### Non-font tokens

The rest of the tokens categories besides font are flatter. The `space` and
`size` generally share keys, and that space can generally use negative keys as
well.

```tsx
// passed into createTamagui
const tokens = createTokens({
  color: {
    white: '#fff',
    black: '#000',
  },
})
```

You access tokens then by using `$` prefixes in your values. Tamagui knows which
tokens to use based on the style property you use.

```tsx
const App = () => (
  <Text fontSize="$lg" lineHeight="$lg" fontFamily="$mono" color="$white">
    Hello world
  </Text>
)
```

One final note: using tokens with themes. Tokens are considered a "fallback" to
themes, so any values you define in your theme will override the token. The next
section will explain this further.

### Configuring tokens

There are a few settings that control how strict your style values are allowed
to be, which are handled by the `settings` option of `createTamagui`. See the
[type strictness settings](#settings) below.

### Themes

Themes live one level below tokens. Tokens are your variables, where themes use
those tokens to create consistent, generic properties that you then typically
use in shareable components. Themes should generally only deal with colors.

Tamagui components in general expect a set of theme keys to be defined like the
following, but you can deviate if you create your own design system.

```tsx
const light = {
  background: '#fff',
  backgroundHover: tokens.color.gray2,
  backgroundPress: tokens.color.gray4,
  backgroundFocus: tokens.color.gray5,
  borderColor: tokens.color.gray4,
  borderColorHover: tokens.color.gray6,
  borderColorPress: tokens.color.gray12,
  borderColorFocus: tokens.color.gray11,
  color: tokens.color.gray10,
  colorHover: tokens.color.gray9,
  colorPress: tokens.color.gray8,
  colorFocus: tokens.color.gray8,
  shadowColor: tokens.color.grayA4,
  shadowColorHover: tokens.color.grayA6,
  shadowColorPress: tokens.color.grayA8,
  shadowColorFocus: tokens.color.grayA8,
  ...lightColors,
}
```

You don't have to use tokens as your theme values, but if you do they avoid some
overhead. With Tamagui, the idea is that `bg`, `color`, and `borderColor`
represent the "primary" and clearest colors, and `bg2`, `color2` etc get more
subtle.

To see how it works, here's a snippet from `InteractiveFrame` which is the frame
component that's used in `Button`:

```tsx
export const InteractiveFrame = styled(XStack, {
  borderRadius: '$1',
  paddingVertical: '$2',
  paddingHorizontal: '$3',
  backgroundColor: '$background',
  justifyContent: 'center',
  alignItems: 'center',
  cursor: 'pointer',
  flexWrap: 'nowrap',
  flexDirection: 'row',
  flexShrink: 1,

  hoverStyle: {
    backgroundColor: '$backgroundHover',
  },

  pressStyle: {
    backgroundColor: '$backgroundPress',
  },

  // ...
})
```

### Media

For more full docs on media queries, see the
[useMedia docs page](/docs/core/use-media).

### Animations

Choose one of `@tamagui/animations-css` for CSS transition based animations,
`@tamagui/animations-react-native` for React Native Animated based animations,
or `@tamagui/animations-moti` for [Moti](https://moti.fyi) animations. You can
swap them out per-platform as well, so long as you match the keys of the
animations you pass in for each driver.

Add `animations` to `createTamagui`:

```tsx
import { createAnimations } from '@tamagui/animations-react-native'

// pass this exported `animations` to your `createTamagui` call:
export const animations = createAnimations({
    bouncy: {
      damping: 9,
      mass: 0.9,
      stiffness: 150,
    },
    lazy: {
      damping: 18,
      stiffness: 50,
    },
  }),
})
```

If using `@tamagui/animations-moti`, be sure to add `moti` as a dependency as
well.

### Shorthands

Shorthands are defined on `createTamagui`. Here's an example of a partial
shorthands configuration:

```tsx
// the as const ensures types work with the optional `onlyAllowShorthands` option
const shorthands = {
  ac: 'alignContent',
  ai: 'alignItems',
  als: 'alignSelf',
  bblr: 'borderBottomLeftRadius',
  bbrr: 'borderBottomRightRadius',
  bg: 'backgroundColor',
  br: 'borderRadius',
  btlr: 'borderTopLeftRadius',
  btrr: 'borderTopRightRadius',
  f: 'flex',
  // ...
} as const

export default createTamagui({
  shorthands,
})
```

Which will enable usage like:

```tsx
<View br="$myToken" />
```

where `br` expands into `borderRadius`.

<Notice theme="green">
  For a full shorthands setup, see [the @tamagui/shorthands
  source](https://github.com/tamagui/tamagui/blob/master/code/core/shorthands/src/index.ts).
</Notice>

### Settings

You can pass a `settings` object to `createTamagui`:

<PropsTable
  data={[
    {
      name: `disableSSR`,
      type: `boolean`,
      description: `For SSR compatibility on the web, Tamagui will render once with the settings from mediaQueryDefaultActive set for all media queries. Then, it will render again after the initial render using the proper media query values. This is so that the hydrating components will match the server. Setting disableSSR to true will avoid this and instead immediately render using the up to date media state, which is the preferrable behavior for client-side only (SPA) style apps.`,
    },
    {
      name: `defaultFont`,
      required: false,
      type: `string`,
      description: `Map it to the regular key of the font given to createTamagui, so "body" or "heading" make sense as values here. This will ensure this font is the fallback for any views that don't define it.`,
    },
    {
      name: 'mediaQueryDefaultActive',
      required: false,
      type: 'Record<string, boolean>',
      description: `For the first render, determines which media queries are true (useful for SSR).`,
    },
    {
      name: 'cssStyleSeparator',
      required: false,
      type: 'string',
      description: `What's between each generated CSS style rule. Set as newline to more easily debug outputted CSS.`,
    },
    {
      name: 'themeClassNameOnRoot',
      required: false,
      type: 'boolean',
      default: false,
      description: `When using next-themes or anything that does SSR and attaches the theme class to the HTML tag, set this to true to have the proper CSS theme selectors generate`,
    },
    {
      name: 'selectionStyles',
      required: false,
      type: '(theme) => ({ backgroundColor: Variable | string; color: Variable | string })',
      default: false,
      description: `On the web, will generate ::selection styles for text selection.`,
    },
    {
      name: 'shouldAddPrefersColorThemes',
      required: false,
      type: 'boolean',
      default: `true`,
      description: `Generates @media queries based on prefers-color-scheme for you if you have light/dark themes.`,
    },
    {
      name: 'maxDarkLightNesting',
      required: false,
      type: 'number',
      default: 3,
      description: `(Advanced) On the web, tamagui treats "dark" and "light" themes as special and generates extra CSS to avoid having to re-render the entire page. This CSS relies on specificity hacks that multiply by your sub-themes. This prop sets the maximum number of nested dark/light themes you can do. Defaults to 3 for a balance, but can be higher if you nest them deeply.`,
    },
    {
      name: 'onlyAllowShorthands',
      required: false,
      type: 'boolean',
      default: false,
      description: `Will remove the type for the long-form versions of any shorthands you define.`,
    },
    {
      name: `allowedStyleValues`,
      required: false,
      type: 'AllowedStyleValuesSetting',
      description: `Set up allowed values on style props, this is only a type-level validation (see below).`,
    },
    {
      name: `autocompleteSpecificTokens`,
      required: false,
      type: `boolean | 'except-special'`,
      description: `Set up if "specific tokens" ($color.name) are added to the types where tokens are allowed (see below).`,
    },
    {
      name: 'mediaPropOrder',
      required: false,
      type: 'boolean',
      default: false,
      description: `(beta) Will change the behavior of media styles. By default they have a fixed specificity: they always override any $theme- or $platform- styles. With this enabled, media styles will have the same precedence as the theme and platform styles, meaning that the order of the props determines if they override.`,
    },
    {
      name: 'fastSchemeChange',
      required: false,
      type: 'boolean',
      default: false,
      description: `On iOS, this enables a mode where Tamagui returns color values using DynamicColorIOS significantly speeding up dark/light re-renders. Note: When enabled you must make sure the defaultTheme you set on TamaguiProvider matches the current system color scheme.`,
    },
    {
      name: 'webContainerType',
      required: false,
      type: 'string',
      default: 'inline-size',
      description: `Only for web, when using group styles Tamagui uses container queries. When using a group on a containing element, Tamagui defaults to setting "container-type" in CSS to "inline-size".`,
    },
  ]}
/>

#### Type strictness

#### `allowedStyleValues`

- `false (default)`: allows any string (or number for styles that accept
  numbers)
- `strict`: only allows tokens for any token-enabled properties
- `strict-web`: same as strict but allows for web-specific tokens like
  auto/inherit
- `somewhat-strict`: allow tokens or:
  - for `space/size`: string% or numbers
  - for `radius`: number
  - for `zIndex`: number
  - for `color`: named colors or rgba/hsla strings
- `somewhat-strict-web`: same as somewhat-strict but allows for web-specific
  tokens

```tsx
type AllowedValueSetting =
  | boolean
  | 'strict'
  | 'somewhat-strict'
  | 'strict-web'
  | 'somewhat-strict-web'

type AllowedStyleValuesSetting =
  | AllowedValueSetting
  | {
      space?: AllowedValueSetting
      size?: AllowedValueSetting
      radius?: AllowedValueSetting
      zIndex?: AllowedValueSetting
      color?: AllowedValueSetting
    }
```

#### `autocompleteSpecificTokens`

The VSCode autocomplete puts specific tokens above the regular ones, which leads
to worse DX. If true this setting removes the specific token from types for the
defined categories.

If set to `except-special`, specific tokens will autocomplete only if they don't
normally use one of the special token groups: space, size, radius, zIndex,
color.
