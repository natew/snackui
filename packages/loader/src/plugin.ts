import { TamaguiOptions, watchTamaguiConfig } from '@tamagui/static'
import type { Compiler, RuleSetRule } from 'webpack'

type PluginOptions = TamaguiOptions & {
  isServer?: boolean
  exclude?: RuleSetRule['exclude']
  test?: RuleSetRule['test']
  jsLoader?: any
  disableEsbuildLoader?: boolean
  disableModuleJSXEntry?: boolean
  disableWatchConfig?: boolean
}

function prettifyWebpackConfig(config) {
  return require('prettyjson').render(config)
}

export class TamaguiPlugin {
  pluginName = 'TamaguiPlugin'

  constructor(
    public options: PluginOptions = {
      components: ['@tamagui/core'],
    }
  ) {}

  apply(compiler: Compiler) {
    if (!this.options.disableWatchConfig) watchTamaguiConfig(this.options)

    // mark as side effect
    compiler.hooks.normalModuleFactory.tap(this.pluginName, (nmf) => {
      nmf.hooks.createModule.tap(
        this.pluginName,
        // @ts-expect-error CreateData is typed as 'object'...
        (createData: { matchResource?: string; settings: { sideEffects?: boolean } }) => {
          if (createData.matchResource?.endsWith('.tamagui.css')) {
            createData.settings.sideEffects = true
          }
        }
      )
    })

    compiler.options.resolve.extensions = [
      ...new Set([
        '.web.tsx',
        '.web.ts',
        '.web.js',
        '.ts',
        '.tsx',
        '.js',
        ...(compiler.options.resolve.extensions || []),
      ]),
    ]

    // look for compiled js with jsx intact as specified by module:jsx
    const mainFields = compiler.options.resolve.mainFields
    if (mainFields) {
      compiler.options.resolve.mainFields = Array.isArray(mainFields)
        ? mainFields
        : [mainFields]
      if (!this.options.disableModuleJSXEntry) mainFields.unshift('module:jsx')
    }

    if (!compiler.options.module) {
      return
    }

    const { jsLoader } = this.options

    const existing = compiler.options.module.rules as any[]

    const rules =
      (existing.find((x) => (typeof x === 'object' && 'oneOf' in x ? x : null))
        ?.oneOf as any[]) ?? existing

    const nextJsRules = rules.findIndex(
      (x) => x?.use && x.use.loader === 'next-swc-loader' && x.issuerLayer !== 'api'
    )

    const esbuildLoader = {
      loader: require.resolve('esbuild-loader'),
      options: {
        target: 'es2021',
        keepNames: true,
        loader: 'tsx',
        tsconfigRaw: {
          module: this.options.isServer ? 'commonjs' : 'esnext',
          isolatedModules: true,
          resolveJsonModule: true,
        },
      },
    }

    const tamaguiLoader = {
      loader: require.resolve('tamagui-loader'),
      options: {
        ...this.options,
      },
    }

    if (nextJsRules === -1) {
      existing.push({
        test: /\/jsx\/.*\.m?[jt]sx?$/,
        exclude: this.options.exclude,
        resolve: {
          fullySpecified: false,
        },
        use: [esbuildLoader],
      })

      // app dir or not next.js
      existing.push({
        test: this.options.test ?? /\.m?[jt]sx?$/,
        exclude: this.options.exclude,
        resolve: {
          fullySpecified: false,
        },
        use: [tamaguiLoader],
      })
    } else if (!this.options.disableEsbuildLoader) {
      const startIndex = nextJsRules ? nextJsRules + 1 : 0
      const existingLoader = nextJsRules ? rules[startIndex] : undefined

      rules.splice(startIndex, 0, {
        test: this.options.test ?? /\.m?[jt]sx?$/,
        exclude: this.options.exclude,
        resolve: {
          fullySpecified: false,
        },
        use: [
          ...(jsLoader ? [jsLoader] : []),
          ...(existingLoader && nextJsRules ? [].concat(existingLoader.use) : []),
          ...(!(jsLoader || existingLoader) ? [esbuildLoader] : []),
          tamaguiLoader,
        ],
      })
    }
  }
}
