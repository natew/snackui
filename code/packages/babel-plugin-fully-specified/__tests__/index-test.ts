import { describe, expect, test, it } from '@jest/globals'
import { type TransformOptions, transform } from '@babel/core'

import plugin, { type FullySpecifiedOptions } from '../src'

/** A test helper for calling Babel transform with the plugin to test and some default options. */
function getTransformResult(
  input,
  {
    transformOptions,
    pluginOptions,
  }: { transformOptions?: TransformOptions; pluginOptions?: FullySpecifiedOptions } = {}
) {
  return transform(input, {
    filename: 'myFile.js',
    configFile: false,
    ...transformOptions,
    plugins: [[plugin, { ensureFileExists: false, ...pluginOptions }]],
  })
}

describe('local imports', () => {
  test('named import', () => {
    const example1 = "import { foo } from './foo'"
    expect(getTransformResult(example1)?.code).toBe('import { foo } from "./foo.js";')

    const example2 = "import { foo as bar } from './foo'"
    expect(getTransformResult(example2)?.code).toBe(
      'import { foo as bar } from "./foo.js";'
    )
  })

  test('default import', () => {
    const example1 = "import defaultFoo from './foo'"
    expect(getTransformResult(example1)?.code).toBe('import defaultFoo from "./foo.js";')

    const example2 = "import defaultFoo, { foo } from './foo'"
    expect(getTransformResult(example2)?.code).toBe(
      'import defaultFoo, { foo } from "./foo.js";'
    )
  })

  test('namespace import', () => {
    const example = "import * as fooModule from './foo'"
    expect(getTransformResult(example)?.code).toBe(
      'import * as fooModule from "./foo.js";'
    )
  })

  test('side effects only import', () => {
    const example = "import './foo'"
    expect(getTransformResult(example)?.code).toBe('import "./foo.js";')
  })
})

describe('local re-exports', () => {
  test('named', () => {
    const example1 = "export { foo } from './foo'"
    expect(getTransformResult(example1)?.code).toBe('export { foo } from "./foo.js";')

    const example2 = "export { foo as bar } from './foo'"
    expect(getTransformResult(example2)?.code).toBe(
      'export { foo as bar } from "./foo.js";'
    )
  })

  test('default', () => {
    const example1 = "export { default as fooDefault } from './foo'"
    expect(getTransformResult(example1)?.code).toBe(
      'export { default as fooDefault } from "./foo.js";'
    )
  })

  test('namespace', () => {
    const example1 = "export * as fooModule from './foo'"
    expect(getTransformResult(example1)?.code).toBe(
      'export * as fooModule from "./foo.js";'
    )

    const example2 = "export * from './foo'"
    expect(getTransformResult(example2)?.code).toBe('export * from "./foo.js";')
  })
})
