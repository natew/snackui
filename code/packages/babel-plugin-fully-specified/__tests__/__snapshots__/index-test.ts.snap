// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`transforming actual files test 1`] = `
"import foo, { bar } from "./modules/module.mjs";
import { foo as foo2 } from "./modules/module.mjs";
import * as someModule from "./modules/module.mjs";
import { foo as cjsFoo } from "./modules/cjs-module.mjs";
import { transform } from '@babel/core';

// Will be transformed if \`includePackages\` includes \`@babel/core\`.
import { transformFile } from "@babel/core/src/transform-file-browser.js";"
`;
